tools/webhook-apicoverage/webhook/apicoverage_recorder.go

```
@@ -0,0 +1,118 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package webhook
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/knative/pkg/webhook"
+	"github.com/knative/test-infra/tools/webhook-apicoverage/coveragecalculator"
+	"github.com/knative/test-infra/tools/webhook-apicoverage/resourcetree"
+	"github.com/knative/test-infra/tools/webhook-apicoverage/view"
+	"go.uber.org/zap"
+	"io/ioutil"
+	"k8s.io/api/admission/v1beta1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"net/http"
+	"os"
+	"reflect"
+)
+
+var (
+	decoder = serializer.NewCodecFactory(runtime.NewScheme()).UniversalDeserializer()
+)
+
+const (
+	// ResourceQueryParam query param name to provide the resource.
+	ResourceQueryParam = "resource"
+)
+
+// APICoverageRecorder type contains resource tree to record API coverage for resources.
+type APICoverageRecorder struct {
+	Logger *zap.SugaredLogger
+	ResourceForest resourcetree.ResourceForest
+	ResourceMap map[schema.GroupVersionKind]webhook.GenericCRD
+	NodeRules resourcetree.NodeRules
+	FieldRules resourcetree.FieldRules
+	DisplayRules view.DisplayRules
+}
+
+// Init initializes the resources tress for set resources.
+func (a *APICoverageRecorder) Init() {
+	for resourceKind, resourceObj := range a.ResourceMap {
+		a.ResourceForest.AddResourceTree(resourceKind.Kind, reflect.ValueOf(resourceObj).Elem().Type())
+	}
+}
+
+// RecordResourceCoverage updates the resource tree with the request.
+func (a *APICoverageRecorder) RecordResourceCoverage(w http.ResponseWriter, r *http.Request) {
+	var body []byte
+	if data, err := ioutil.ReadAll(r.Body); err == nil {
```

can you comment why ignore error here?

---

Actually have not handled the error cases correctly in this handler. Fixed it here and all places.